[#12][UI] As a user, I can see survey intro - Fetch survey detail

hanam1ni opened the pull request:
## What happened 👀

Describe the big picture of your changes here to communicate to the team why we should accept this pull request. 
 
## Insight 📝

Describe in details how to test the changes, which solution you tried but did not go with, referenced documentation is welcome as well.
 
## Proof Of Work 📹

Show us the implementation: screenshots, gif, etc.

hanam1ni commented on cypress/integration/Home/index.spec.ts:
```diff
@@ -1,7 +1,11 @@
   const scrollSurveyItem = (times = 1) => {
     for (let time = 1; time <= times; time++) {
-      cy.get('[data-testid="survey-list"]').move({ deltaX: -500, deltaY: 100 });
+      cy.get('[data-testid="survey-list"]')
```
Note to reviewers: Fix random CI failing

malparty commented on src/services/survey.ts:
```diff
@@ -23,16 +25,83 @@ export const listSurveys = async (options = {}): Promise<SurveyResponse> => {
+    parseSurveyQuestion(question)
+  );
+
+  [[surveyIntro], questions] = extractQuestionType(questions, 'intro');
+  [[surveyOutro], questions] = extractQuestionType(questions, 'outro');
```
I should first say thanks because I finally took the time to play and learn lodash thanks to this PR :D 

Now, I'm wondering if using `partition` as a filter is not a little bit overkill... 
What about using filters only?

```suggestion
  surveyIntro = questions.filter(q => q.type === 'intro');
  surveyOutro = questions.filter(q => q.type === 'outro');
  questions = questions.filter(q => q.type !== 'intro' && q.type !== 'outro');
```

hanam1ni replied:
@malparty At first, I consider the filter function, but I don't want to iterate the whole question list when extracting the intro and outro. Thinking again, I think it's fine because the question list is not large.

malparty replied:
@hanam1ni  And maybe also because the "partition" helper needs to iterate the whole questions array as well ;-)
By the way, it is very interesting to dive in `lodash` source code: [The partition helper](https://github.com/lodash/lodash/blob/master/partition.js) is using [a custom "reduce"](https://github.com/lodash/lodash/blob/2da024c3b4f9947a48517639de7560457cd4ec6c/.internal/arrayReduce.js#L19) method that finally does a `while` covering all indexes of the array. 😇 

<img width="480" alt="image" src="https://user-images.githubusercontent.com/77609814/175907530-8c9cf241-e11b-4651-91ec-15cf1b700b86.png">

malparty commented on src/services/survey.ts:
```diff
@@ -23,16 +25,83 @@ export const listSurveys = async (options = {}): Promise<SurveyResponse> => {
+  remainingQuestions: SurveyQuestion[]
+] => partition(questions, (question) => question.displayType == displayType);
+
+const sortRecords = (records: object) =>
+  sortBy(records, [({ displayOrder }) => displayOrder]);
```
```suggestion
const sortRecords = (records: object) =>
  records.sort((a, b) => b.displayOrder - a.displayOrder);
```

Does it work with that? If yes, it means you might not need lodash at all 😇

malparty reviewed (changes requested):
Minor suggestions but no blockers ;-)

hoangmirs commented on src/pages/surveys/[id].page.tsx:
```diff
@@ -1,10 +1,32 @@
+const SurveyDetail: NextPage = () => {
   const router = useRouter();
-  const { id } = router.query;
+  const surveyId = router.query.id as string | undefined;
```
I think we can use `getServerSideProps` to get the `surveryId` instead.
You can check it [here](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#when-does-getserversideprops-run).

We can fetch the data from the server side, and pass it directly to the prop of this component.

The advantage of this way is the content can be served faster than fetching from the client-side.

What do you think? 🤔

hanam1ni replied:
@hoangmirs I looked at this before, but with the current approach I use to make an HTTP request to the backend, it doesn't support `getServerSideProps`. Because the HTTP request requires a user token which stored in local storage, it can't be accessed on the server-side. 
I need to refactor passing user token as HttpOnly cookie and create a proxy for all the API calls. So I plan to finish the whole feature first after that I can revisit this refactoring if I have time.

hanam1ni commented:
Removed Lodash and addressed all your comments here [9ebc34d](https://github.com/hanam1ni/react-surveyor/pull/76/commits/9ebc34df23471a83096bd01839ea1b6ab6b9e7e3)

malparty reviewed (approved):
All good to me 🙌
